### PR TITLE
*: add Copy() to transaction.Transaction and payload.P2PNotaryRequest

### LINKED ISSUE #3288: Add Copy() to transaction.Transaction and payload.P2PNotaryRequest
## Is your feature request related to a problem? Please describe.
I'm always frustrated when I need to make a deep copy of some structure, but there is no method provided for that. Copying transactions and notary requests happens often enough.

## Describe the solution you'd like
Add a method that makes a deep copy of all fields (and resets size/hash caches!).

## Describe alternatives you've considered
Keep copy-pasting local `safeCopy()` and alike functions.

## Additional context
https://github.com/nspcc-dev/neofs-node/pull/2715

### PR DESCRIPTION
**Pull request by AliceInHunterland**
Close #3288

### CODE REVIEW

File: pkg/core/transaction/transaction.go [hunk 1]
```diff
@@ -468,3 +468,32 @@ func (t *Transaction) ToStackItem() stackitem.Item {
...
+	}
+	if t.Scripts != nil {
+		txCopy.Scripts = make([]Witness, len(t.Scripts))
+		for i, script := range t.Scripts {
+			txCopy.Scripts[i] = *script.Copy()
+		}
+	}
+	if t.Script != nil {
+		txCopy.Script = make([]byte, len(t.Script))
+		copy(txCopy.Script, t.Script)
```
Comment (roman-khimov) [inline pkg/core/transaction/transaction.go:492]: https://pkg.go.dev/bytes#Clone

File: pkg/core/transaction/witness.go [hunk 2]
```diff
@@ -38,3 +38,20 @@ func (w *Witness) EncodeBinary(bw *io.BinWriter) {
...
 	return hash.Hash160(w.VerificationScript)
 }
+
+// Copy creates a deep copy of the Witness.
+func (w *Witness) Copy() *Witness {
+	wCopy := &Witness{}
+
+	if w.InvocationScript != nil {
+		wCopy.InvocationScript = make([]byte, len(w.InvocationScript))
+		copy(wCopy.InvocationScript, w.InvocationScript)
```
Comment (roman-khimov) [inline pkg/core/transaction/witness.go:48]: https://pkg.go.dev/bytes#Clone

File: pkg/core/transaction/witness.go [hunk 3]
```diff
@@ -38,3 +38,20 @@ func (w *Witness) EncodeBinary(bw *io.BinWriter) {
...
+	}
+
+	if w.VerificationScript != nil {
+		wCopy.VerificationScript = make([]byte, len(w.VerificationScript))
+		copy(wCopy.VerificationScript, w.VerificationScript)
```
Comment (roman-khimov) [inline pkg/core/transaction/witness.go:53]: https://pkg.go.dev/bytes#Clone

It handles `nil` as well, btw.

File: pkg/core/transaction/transaction_test.go [hunk 4]
```diff
@@ -321,3 +321,28 @@ func BenchmarkTxHash(b *testing.B) {
...
+	origTx.NetworkFee = 2000
+	origTx.Signers = []Signer{{Account: random.Uint160()}}
+	origTx.Scripts = []Witness{
+		{
+			InvocationScript:   []byte{0x04, 0x05},
+			VerificationScript: []byte{0x06, 0x07},
+		},
+	}
+
+	copyTx := origTx.Copy()
```
Comment (roman-khimov) [inline pkg/core/transaction/transaction_test.go:336]: Not all branches are covered here.

File: pkg/core/transaction/transaction_test.go [hunk 5]
```diff
@@ -321,3 +321,28 @@ func BenchmarkTxHash(b *testing.B) {
...
+
+	require.Equal(t, origTx, copyTx)
+
+	copyTx.NetworkFee = 3000
+	copyTx.Scripts[0].InvocationScript[0] = 0x08
+
```
Comment (roman-khimov) [inline pkg/core/transaction/transaction_test.go:342]: Check hashes after the change as well.

File: pkg/network/payload/notary_request.go [hunk 6]
```diff
@@ -136,3 +136,22 @@ func (r *P2PNotaryRequest) isValid() error {
 	}
 	return nil
 }
+
+// Copy creates a deep copy of the P2PNotaryRequest.
+func (r *P2PNotaryRequest) Copy() *P2PNotaryRequest {
+	var mainTxCopy, fallbackTxCopy *transaction.Transaction
+	if r.MainTransaction != nil {
+		mainTxCopy = r.MainTransaction.Copy()
```
Comment (roman-khimov) [inline pkg/network/payload/notary_request.go:144]: You can handle `nil` in `Copy()` as well.

File: pkg/core/transaction/transaction.go [hunk 7]
```diff
@@ -468,3 +468,32 @@ func (t *Transaction) ToStackItem() stackitem.Item {
...
+	}
+
+	txCopy.hashed = false
+	txCopy.size = 0
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/core/transaction/transaction.go:496]: Then let's just add a couple of words about it to the method documentation.

File: pkg/core/transaction/transaction.go [hunk 8]
```diff
@@ -468,3 +469,33 @@ func (t *Transaction) ToStackItem() stackitem.Item {
...
+	if t.Scripts != nil {
+		txCopy.Scripts = make([]Witness, len(t.Scripts))
+		for i, script := range t.Scripts {
+			txCopy.Scripts[i] = *script.Copy()
+		}
+	}
+	txCopy.Script = bytes.Clone(t.Script)
+
+	txCopy.hashed = false
+	txCopy.size = 0
```
Comment (AnnaShaleva) [inline pkg/core/transaction/transaction.go:498]: Technically, it's enough, but I'd prefer to keep the resulting transaction consistent, and thus, `hash` should also be reset.

File: pkg/network/payload/notary_request.go [hunk 9]
```diff
@@ -136,3 +136,20 @@ func (r *P2PNotaryRequest) isValid() error {
...
 	return nil
 }
+
+// Copy creates a deep copy of the P2PNotaryRequest.
+func (r *P2PNotaryRequest) Copy() *P2PNotaryRequest {
+	var mainTxCopy, fallbackTxCopy *transaction.Transaction
+	mainTxCopy = r.MainTransaction.Copy()
+
+	fallbackTxCopy = r.FallbackTransaction.Copy()
+
```
Comment (AnnaShaleva) [inline pkg/network/payload/notary_request.go:146]: A lot of useless empty lines. And `mainTxCopy`, `fallbackTxCopy` vars are not needed, initialize the `&P2PNotaryRequest` fields in-place.

File: pkg/network/payload/notary_request.go [hunk 10]
```diff
@@ -136,3 +136,20 @@ func (r *P2PNotaryRequest) isValid() error {
...
+	witnessCopy := r.Witness.Copy()
+
+	return &P2PNotaryRequest{
+		MainTransaction:     mainTxCopy,
+		FallbackTransaction: fallbackTxCopy,
+		Witness:             *witnessCopy,
+		hash:                r.hash,
```
Comment (AnnaShaleva) [inline pkg/network/payload/notary_request.go:153]: Hash initialization is not needed then? To keep this helper consistent with transaction's `Copy()`.

File: pkg/core/transaction/witness.go [hunk 11]
```diff
@@ -38,3 +40,11 @@ func (w *Witness) EncodeBinary(bw *io.BinWriter) {
 func (w Witness) ScriptHash() util.Uint160 {
 	return hash.Hash160(w.VerificationScript)
 }
+
+// Copy creates a deep copy of the Witness.
+func (w *Witness) Copy() *Witness {
+	wCopy := &Witness{}
+	wCopy.InvocationScript = bytes.Clone(w.InvocationScript)
+	wCopy.VerificationScript = bytes.Clone(w.VerificationScript)
+	return wCopy
```
Comment (AnnaShaleva) [inline pkg/core/transaction/witness.go:49]:
```suggestion
	return &Witness{
		InvocationScript: bytes.Clone(w.InvocationScript),
		VerificationScript: bytes.Clone(w.VerificationScript),
	}
```

File: pkg/services/notary/notary.go [hunk 12]
```diff
@@ -345,21 +345,6 @@ func (n *Notary) OnNewRequest(payload *payload.P2PNotaryRequest) {
 	}
 }
 
-// safeCopy creates a copy of provided transaction by dereferencing it and creating
-// fresh witnesses so that the tx's witnesses may be modified without affecting the
-// copy's ones.
-func safeCopy(tx *transaction.Transaction) *transaction.Transaction {
```
Comment (AnnaShaleva) [inline pkg/services/notary/notary.go:351]: IIRC, it's not the only place where `Copy()` should be used.

File: pkg/services/notary/notary.go [hunk 13]
```diff
@@ -285,7 +285,7 @@ func (n *Notary) OnNewRequest(payload *payload.P2PNotaryRequest) {
 	// size won't be changed after finalisation, the witness bytes changes may
 	// affect the other users of notary pool and cause race. Avoid this by making
 	// the copy.
-	r.fallbacks = append(r.fallbacks, safeCopy(payload.FallbackTransaction))
+	r.fallbacks = append(r.fallbacks, payload.FallbackTransaction.Copy())
```
Comment (AnnaShaleva) [inline pkg/services/notary/notary.go:288]: Should be done in a separate commit.

File: pkg/services/notary/notary.go (hunk 12 above)
Comment (AliceInHunterland) [reply pkg/services/notary/notary.go:351]: checked all `copy` places with this word. Do you have any other suggestions to find its usage?

File: pkg/core/transaction/transaction.go [hunk 14]
```diff
@@ -468,3 +469,33 @@ func (t *Transaction) ToStackItem() stackitem.Item {
...
+// 'hashed' and 'size' are reset to ensure the copy can be modified independently of the original.
+func (t *Transaction) Copy() *Transaction {
+	if t == nil {
+		return nil
+	}
+	txCopy := *t
+
+	if t.Attributes != nil {
+		txCopy.Attributes = make([]Attribute, len(t.Attributes))
+		copy(txCopy.Attributes, t.Attributes)
```
Comment (AnnaShaleva) [inline pkg/core/transaction/transaction.go:483]: This won't work, we need `Copy()` to be defined on `Attribute` since attribute value is a pointer.

@roman-khimov, agree? Since we're declaring that `Copy()` is a full deep copy.

File: pkg/core/transaction/transaction_test.go (hunk 4 above)
Comment (AnnaShaleva) [reply pkg/core/transaction/transaction_test.go:337]: Still not resolved. We need transaction with all fields filled, including attributes, signers with scopes, rules, etc.

File: pkg/core/transaction/witness.go [hunk 15]
```diff
@@ -38,3 +40,11 @@ func (w *Witness) EncodeBinary(bw *io.BinWriter) {
 func (w Witness) ScriptHash() util.Uint160 {
 	return hash.Hash160(w.VerificationScript)
 }
+
+// Copy creates a deep copy of the Witness.
+func (w *Witness) Copy() *Witness {
```
Comment (AnnaShaleva) [inline pkg/core/transaction/witness.go:45]: We need to define it on structure instead of a pointer, and return a structure instead of pointer so that the result can be directly used to initialize new tx's witness without dereference.

File: pkg/network/payload/notary_request_test.go [hunk 16]
```diff
@@ -186,3 +186,53 @@ func TestNotaryRequestBytesFromBytes(t *testing.T) {
...
+
+func TestP2PNotaryRequest_Copy(t *testing.T) {
+	orig := &P2PNotaryRequest{
+		MainTransaction: &transaction.Transaction{
+			Version:    1,
+			SystemFee:  100,
+			NetworkFee: 50,
+			Script:     []byte{0, 1, 2},
+			Signers:    []transaction.Signer{{Account: util.Uint160{1, 5, 9}}},
+			Attributes: []transaction.Attribute{{Type: transaction.HighPriority}},
```
Comment (AnnaShaleva) [inline pkg/network/payload/notary_request_test.go:198]: Need some attribute with value.

File: pkg/network/payload/notary_request_test.go [hunk 17]
```diff
@@ -186,3 +186,53 @@ func TestNotaryRequestBytesFromBytes(t *testing.T) {
...
 }
+
+func TestP2PNotaryRequest_Copy(t *testing.T) {
+	orig := &P2PNotaryRequest{
+		MainTransaction: &transaction.Transaction{
+			Version:    1,
+			SystemFee:  100,
+			NetworkFee: 50,
+			Script:     []byte{0, 1, 2},
+			Signers:    []transaction.Signer{{Account: util.Uint160{1, 5, 9}}},
```
Comment (AnnaShaleva) [inline pkg/network/payload/notary_request_test.go:197]: Need more complicated signer.

File: pkg/services/notary/notary.go (hunk 12 above)
Comment (AnnaShaleva) [reply pkg/services/notary/notary.go:351]: I thought about Oracle service, but it doesn't need this new `Copy()`, so OK for now.

File: pkg/network/payload/notary_request.go [hunk 18]
```diff
@@ -136,3 +136,12 @@ func (r *P2PNotaryRequest) isValid() error {
 	}
 	return nil
 }
+
+// Copy creates a deep copy of the P2PNotaryRequest.
+func (r *P2PNotaryRequest) Copy() *P2PNotaryRequest {
```
Comment (AnnaShaleva) [inline pkg/network/payload/notary_request.go:141]: A couple of words about cached values are needed in the doc, like for transaction's `Copy()`.

**Comment by AliceInHunterland** [comment]
@AnnaShaleva Let's decide if the current copy implementation is okay before writing tests for all new copy() methods?

File: pkg/core/transaction/witness_condition_test.go [hunk 19]
```diff
@@ -36,6 +36,10 @@ func (c InvalidCondition) ToStackItem() stackitem.Item {
 	panic("invalid")
 }
 
+func (c InvalidCondition) Copy() WitnessCondition {
```
Comment (AnnaShaleva) [inline pkg/core/transaction/witness_condition_test.go:39]: Missing comment.

File: pkg/crypto/keys/publickey.go [hunk 20]
```diff
@@ -406,3 +408,19 @@ func (p *PublicKey) UnmarshalYAML(unmarshal func(any) error) error {
...
+	if p.X != nil {
+		copiedX := new(big.Int).Set(p.X)
+		copied.X = copiedX
+	}
+	if p.Y != nil {
+		copiedY := new(big.Int).Set(p.Y)
+		copied.Y = copiedY
+	}
+	return &copied
+}
```
Comment (AnnaShaleva) [inline pkg/crypto/keys/publickey.go:426]: OK, I think we need @roman-khimov's opinion on this. Do we need such deep copies?

File: pkg/core/transaction/signer.go [hunk 21]
```diff
@@ -86,3 +86,26 @@ func SignersToStackItem(signers []Signer) stackitem.Item {
...
+
+// Copy creates a deep copy of the Signer.
+func (c *Signer) Copy() *Signer {
+	if c == nil {
+		return nil
+	}
+	signerCopy := *c
+	if c.AllowedContracts != nil {
+		signerCopy.AllowedContracts = make([]util.Uint160, len(c.AllowedContracts))
+		copy(signerCopy.AllowedContracts, c.AllowedContracts)
```
Comment (AnnaShaleva) [inline pkg/core/transaction/signer.go:98]: @roman-khimov, for `util.Uint160` the question is the same, do we need the *deep* copy of every signer's hash?

File: pkg/crypto/keys/publickey.go (hunk 20 above)
Comment (AnnaShaleva) [reply pkg/crypto/keys/publickey.go:426]: Discussed, we don't need this for `PublicKey`, but need to add documentation about this behaviour.

File: pkg/core/transaction/attribute_test.go [hunk 22]
```diff
@@ -177,3 +178,35 @@ func TestAttribute_MarshalJSON(t *testing.T) {
...
+		{Type: HighPriority},
+		{Type: OracleResponseT, Value: &OracleResponse{ID: 123, Code: Success, Result: []byte{1, 2, 3}}},
+		{Type: NotValidBeforeT, Value: &NotValidBefore{Height: 123}},
+		{Type: ConflictsT, Value: &Conflicts{Hash: random.Uint256()}},
+		{Type: NotaryAssistedT, Value: &NotaryAssisted{NKeys: 3}},
+	}
+
+	for _, original := range originals {
+		copied := original.Copy()
+		require.NotNil(t, copied, "Copy should not be nil")
```
Comment (AnnaShaleva) [inline pkg/core/transaction/attribute_test.go:193]: Useless error message, remove it, it will be clear from the failing test log.

File: pkg/core/transaction/transaction_test.go (hunk 4 above)
Comment (AliceInHunterland) [reply pkg/core/transaction/transaction_test.go:355]: is it resolved?

File: pkg/core/transaction/attribute.go [hunk 23]
```diff
@@ -8,17 +8,21 @@ import (
...
 
+// AttrValue represents a Transaction Attribute value.
+type AttrValue interface {
+	io.Serializable
+	// toJSONMap is used for embedded json struct marshalling.
+	// Anonymous interface fields are not considered anonymous by
+	// json lib and marshaling Value together with type makes code
+	// harder to follow.
+	toJSONMap(map[string]any)
+	Copy() AttrValue
```
Comment (AnnaShaleva) [inline pkg/core/transaction/attribute.go:19]: Belongs to the next commit.

Comment (AnnaShaleva) [inline pkg/core/transaction/attribute.go:19]: Deserves a comment.

File: pkg/core/transaction/attribute.go [hunk 24]
```diff
@@ -8,17 +8,21 @@ import (
 	"github.com/nspcc-dev/neo-go/pkg/io"
 )
 
+// AttrValue represents a Transaction Attribute value.
+type AttrValue interface {
```
Comment (AnnaShaleva) [inline pkg/core/transaction/attribute.go:12]: Fix the commit message, it's not refactor. It's "move transaction Attribute value to a designated interface".

File: pkg/core/transaction/witness_condition.go [hunk 25]
```diff
@@ -310,6 +330,13 @@ func (c *ConditionOr) ToStackItem() stackitem.Item {
 	return condToStackItem(c.Type(), []WitnessCondition(*c))
 }
 
+// Copy implements the WitnessCondition interface and returns a deep copy of the condition.
+func (c *ConditionOr) Copy() WitnessCondition {
+	conds := make([]WitnessCondition, len(*c))
+	copy(*c, conds)
```
Comment (AnnaShaleva) [inline pkg/core/transaction/witness_condition.go:336]: No way, you need a `Copy()` for every condition.

File: pkg/core/transaction/witness_condition.go [hunk 26]
```diff
@@ -264,6 +277,13 @@ func (c *ConditionAnd) ToStackItem() stackitem.Item {
 	return condToStackItem(c.Type(), []WitnessCondition(*c))
 }
 
+// Copy implements the WitnessCondition interface and returns a deep copy of the condition.
+func (c *ConditionAnd) Copy() WitnessCondition {
+	conds := make([]WitnessCondition, len(*c))
+	copy(*c, conds)
```
Comment (AnnaShaleva) [inline pkg/core/transaction/witness_condition.go:283]: Ditto, `Copy()` for every condition.

File: pkg/core/transaction/witness_condition_test.go [hunk 27]
```diff
@@ -347,3 +352,52 @@ func TestWitnessConditionMatch(t *testing.T) {
...
+			require.NotEqual(t, c, cond.(*ConditionCalledByContract))
+		case *ConditionCalledByGroup:
+			c = (*ConditionCalledByGroup)(pk.PublicKey())
+			require.NotSame(t, c, cond.(*ConditionCalledByGroup))
+		case *ConditionNot:
+			require.NotSame(t, copied, cond)
+		case *ConditionAnd:
+			require.NotSame(t, copied, cond)
+		case *ConditionOr:
+			require.NotSame(t, copied, cond)
```
Comment (AnnaShaleva) [inline pkg/core/transaction/witness_condition_test.go:398]: For the nested we need to check not only this but also that if you change the child, then copy is not changed.

File: pkg/core/transaction/witness_test.go [hunk 28]
```diff
@@ -36,3 +36,16 @@ func TestWitnessSerDes(t *testing.T) {
 	require.Error(t, testserdes.DecodeBinary(bin1, exp))
 	require.Error(t, testserdes.DecodeBinary(bin2, exp))
 }
+
+func TestWitnessCopy(t *testing.T) {
+	original := &Witness{
+		InvocationScript:   make([]byte, 64),
+		VerificationScript: make([]byte, 32),
```
Comment (AnnaShaleva) [inline pkg/core/transaction/witness_test.go:43]: Make it non-nil. Just `[]byte{1, 2, 3}`.

File: pkg/core/transaction/witness_test.go [hunk 29]
```diff
@@ -36,3 +36,16 @@ func TestWitnessSerDes(t *testing.T) {
...
+	}
+
+	copyWitness := original.Copy()
+	require.Equal(t, *original, copyWitness)
+
+	copyWitness.VerificationScript = make([]byte, MaxVerificationScript+1)
```
Comment (AnnaShaleva) [inline pkg/core/transaction/witness_test.go:49]: And here instead just change some byte of the original witness, like `original.InvocationScript[0] = 0x05`.

File: pkg/core/transaction/witness_test.go [hunk 30]
```diff
@@ -36,3 +36,16 @@ func TestWitnessSerDes(t *testing.T) {
...
 	require.Error(t, testserdes.DecodeBinary(bin2, exp))
 }
+
+func TestWitnessCopy(t *testing.T) {
+	original := &Witness{
+		InvocationScript:   make([]byte, 64),
+		VerificationScript: make([]byte, 32),
+	}
+
+	copyWitness := original.Copy()
```
Comment (AnnaShaleva) [inline pkg/core/transaction/witness_test.go:46]: Here and everywhere else, just `cp`, it's shorter and easy to read.

File: pkg/core/transaction/witness_condition.go [hunk 31]
```diff
@@ -310,6 +333,15 @@ func (c *ConditionOr) ToStackItem() stackitem.Item {
 	return condToStackItem(c.Type(), []WitnessCondition(*c))
 }
 
+// Copy implements the WitnessCondition interface and returns a deep copy of the condition.
+func (c *ConditionOr) Copy() WitnessCondition {
+	copiedConditions := make(ConditionOr, len(*c))
```
Comment (AnnaShaleva) [inline pkg/core/transaction/witness_condition.go:338]: s/copiedConditions/cp

File: pkg/core/transaction/witness_condition.go [hunk 32]
```diff
@@ -264,6 +278,15 @@ func (c *ConditionAnd) ToStackItem() stackitem.Item {
 	return condToStackItem(c.Type(), []WitnessCondition(*c))
 }
 
+// Copy implements the WitnessCondition interface and returns a deep copy of the condition.
+func (c *ConditionAnd) Copy() WitnessCondition {
+	copiedConditions := make(ConditionAnd, len(*c))
```
Comment (AnnaShaleva) [inline pkg/core/transaction/witness_condition.go:283]: s/copiedConditions/cp

File: pkg/core/transaction/signer.go [hunk 33]
```diff
@@ -86,3 +86,24 @@ func SignersToStackItem(signers []Signer) stackitem.Item {
 	}
 	return stackitem.NewArray(res)
 }
+
+// Copy creates a deep copy of the Signer.
+func (c *Signer) Copy() *Signer {
+	if c == nil {
+		return nil
+	}
+	signerCopy := *c
```
Comment (AnnaShaleva) [inline pkg/core/transaction/signer.go:95]: c/signerCopy/cp

File: pkg/core/transaction/attribute.go [hunk 34]
```diff
@@ -110,3 +112,17 @@ func (attr *Attribute) UnmarshalJSON(data []byte) error {
 	}
 	return json.Unmarshal(data, attr.Value)
 }
+
+// Copy creates a deep copy of the Attribute.
+func (attr *Attribute) Copy() *Attribute {
+	if attr == nil {
+		return nil
+	}
+	copiedAttr := &Attribute{
```
Comment (AnnaShaleva) [inline pkg/core/transaction/attribute.go:121]: s/copiedAttr/cp

File: pkg/services/notary/notary.go [hunk 35]
```diff
@@ -272,7 +272,7 @@ func (n *Notary) OnNewRequest(payload *payload.P2PNotaryRequest) {
 		// Avoid changes in the main transaction witnesses got from the notary request pool to
 		// keep the pooled tx valid. We will update its copy => the copy's size will be changed.
 		r = &request{
-			main:              safeCopy(payload.MainTransaction),
+			main:              payload.MainTransaction.Copy(),
```
Comment (AnnaShaleva) [inline pkg/services/notary/notary.go:275]: > notary: refactor Copy() of transaction.Transaction 

Change the commit message to something more specific. It's not the `Copy()`'s method refactoring as it can be judged from the commit message. It's something like `notary: reuse (*Transaction).Copy where possible` or `notary: get rid of custom safeCopy implementation` or something like that.

**Review by AnnaShaleva — COMMENTED** [review]
Style comments, the logic LGTM.

File: pkg/core/transaction/transaction.go [hunk 36]
```diff
@@ -468,3 +469,36 @@ func (t *Transaction) ToStackItem() stackitem.Item {
...
 	})
 }
+
+// Copy creates a deep copy of the Transaction, including all slice fields. Cached values like
+// 'hashed' and 'size' are reset to ensure the copy can be modified independently of the original.
+func (t *Transaction) Copy() *Transaction {
+	if t == nil {
+		return nil
+	}
+	txCopy := *t
```
Comment (AnnaShaleva) [inline pkg/core/transaction/transaction.go:479]: s/txCopy/cp

File: pkg/crypto/keys/publickey.go [hunk 37]
```diff
@@ -78,8 +78,12 @@ func (keys PublicKeys) Contains(pKey *PublicKey) bool {
 	return false
 }
 
-// Copy returns a copy of keys.
+// Copy returns a shallow copy of the PublicKeys slice. It creates a new slice with the same elements,
+// but does not perform a deep copy of the elements themselves.
 func (keys PublicKeys) Copy() PublicKeys {
+	if keys == nil {
+		return nil
+	}
```
Comment (AnnaShaleva) [inline pkg/crypto/keys/publickey.go:86]: May be moved to a separate commit (along with the test) before the main one since it's a significant change in the old frequently-used API, but it's not critical.

**Review by AnnaShaleva — APPROVED** [review]
Nice!

**Comment by AnnaShaleva** [comment]
@roman-khimov, ready for review.

File: pkg/core/transaction/signer.go [hunk 38]
```diff
@@ -15,11 +15,11 @@ const maxSubitems = 16
...
 type Signer struct {
-	Account          util.Uint160      `json:"account"`
-	Scopes           WitnessScope      `json:"scopes"`
-	AllowedContracts []util.Uint160    `json:"allowedcontracts,omitempty"`
-	AllowedGroups    []*keys.PublicKey `json:"allowedgroups,omitempty"`
-	Rules            []WitnessRule     `json:"rules,omitempty"`
+	Account          util.Uint160    `json:"account"`
+	Scopes           WitnessScope    `json:"scopes"`
+	AllowedContracts []util.Uint160  `json:"allowedcontracts,omitempty"`
+	AllowedGroups    keys.PublicKeys `json:"allowedgroups,omitempty"`
```
Comment (roman-khimov) [inline pkg/core/transaction/signer.go:21]: Do you really need this change? It can break some initializations in existing code.

Although I've tried to search for uses of it and failed to find any... Old node code used it. Maybe @fyfyrchik can tell if it's still in use somewhere.

File: pkg/core/transaction/signer.go [hunk 39]
```diff
@@ -86,3 +86,24 @@ func SignersToStackItem(signers []Signer) stackitem.Item {
...
+func (c *Signer) Copy() *Signer {
+	if c == nil {
+		return nil
+	}
+	cp := *c
+	if c.AllowedContracts != nil {
+		cp.AllowedContracts = make([]util.Uint160, len(c.AllowedContracts))
+		copy(cp.AllowedContracts, c.AllowedContracts)
+	}
+	cp.AllowedGroups = c.AllowedGroups.Copy()
```
Comment (roman-khimov) [inline pkg/core/transaction/signer.go:100]: Because you can cast here, if needed.

File: pkg/core/transaction/witness_test.go [hunk 40]
```diff
@@ -9,19 +9,19 @@ import (
...
-		InvocationScript:   make([]byte, 64),
-		VerificationScript: make([]byte, 32),
+		InvocationScript:   []byte{1, 2, 3},
+		VerificationScript: []byte{3, 2, 1},
 	}
 	var good2 = &Witness{
-		InvocationScript:   make([]byte, MaxInvocationScript),
-		VerificationScript: make([]byte, MaxVerificationScript),
+		InvocationScript:   []byte{1, 2, 2},
+		VerificationScript: []byte{3, 2, 2},
```
Comment (roman-khimov) [inline pkg/core/transaction/witness_test.go:17]: Why?

File: pkg/core/transaction/signer.go (hunk 38 above)
Comment (AliceInHunterland) [reply pkg/core/transaction/signer.go]: okay, moved back to []*keys.PublicKey and adjusted Copy()

File: pkg/core/transaction/signer.go [hunk 41]
```diff
@@ -86,3 +86,34 @@ func SignersToStackItem(signers []Signer) stackitem.Item {
...
+		return nil
+	}
+	cp := *c
+	if c.AllowedContracts != nil {
+		cp.AllowedContracts = make([]util.Uint160, len(c.AllowedContracts))
+		copy(cp.AllowedContracts, c.AllowedContracts)
+	}
+
+	if c.AllowedGroups != nil {
+		cp.AllowedGroups = make([]*keys.PublicKey, len(c.AllowedGroups))
```
Comment (roman-khimov) [inline pkg/core/transaction/signer.go:102]: Consider this code:
```
--- a/pkg/crypto/keys/publickey_test.go
+++ b/pkg/crypto/keys/publickey_test.go
@@ -40,14 +40,18 @@ func TestEncodeDecodePublicKey(t *testing.T) {
 }
 
 func TestPublicKeys_Copy(t *testing.T) {
-       pubs := make(PublicKeys, 5)
-       for i := range pubs {
+       pubz := make([]*PublicKey, 5)
+       for i := range pubz {
                priv, err := NewPrivateKey()
                require.NoError(t, err)
-               pubs[i] = priv.PublicKey()
+               pubz[i] = priv.PublicKey()
        }
+       pubs := PublicKeys(pubz)
 
        cp := pubs.Copy()
+       var pubx = ([]*PublicKey)(cp)
+       require.Equal(t, pubz, pubx)
+
        priv, err := NewPrivateKey()
        require.NoError(t, err)
        cp[0] = priv.PublicKey()
```
Just cast/Copy/cast again, no need for this.

Comment (roman-khimov) [reply pkg/core/transaction/signer.go:102]: I don't see anything changed here.

Comment (AliceInHunterland) [reply pkg/core/transaction/signer.go:102]: added a separate commit with changing the test. https://github.com/nspcc-dev/neo-go/pull/3407/files#diff-617aae89a8bc5f83ec727408c2b9b72f783683dc1d04acd4b9a67c6cbf2331ae

Comment (roman-khimov) [reply pkg/core/transaction/signer.go:102]: Sorry, I don't care much about this test. It was just an example. What I care about a bit more here is dropping code duplication from `(*Signer).Copy()`. The idea with `PublicKeys.Copy()` reuse is a good one, but my point is that you don't need to change `AllowedGroups` type for it. Just cast/copy/cast and don't do if/make/for/copy. Less code, less problems.

**Review by roman-khimov — APPROVED** [review]

**Comment by AnnaShaleva** [comment]
> Do you really need this change? It can break some initializations in existing code.

That was my concern as far, but I've checked the existing code and it seems that this type change doesn't break anything. Anyway, the current version looks better.

**Review by AnnaShaleva — APPROVED** [review]
